pkg/util/kubernetes/hpa/hpa.go

```diff
@@ -0,0 +1,366 @@
+	"k8s.io/client-go/kubernetes"
+)
+
+//// Covered by test/integration/util/kubernetes/hpa/hpa_test.go
+
+var (
+	expectedHPAType = reflect.TypeOf(v2beta1.HorizontalPodAutoscaler{})
+)
+
+const (
+	datadogHPAConfigMap = "datadog-hpa"
+)
+
+type CustomExternalMetric struct {
+	Name      string            `json:"name"`
+	Labels    map[string]string `json:"labels"`
+	Timestamp int64             `json:"ts"`
+	HpaName   string            `json:"origin"`
+	Value     int64             `json:"value"`
+}
+
+// HPAWatcherClient embeds the API Server client and the configuration to refresh metrics from Datadog and watch the HPA Objects' activities
+type HPAWatcherClient struct {
+	clientSet      kubernetes.Interface
+	readTimeout    time.Duration
+	refreshItl     *time.Ticker
+	pollItl        *time.Ticker
+	externalMaxAge time.Duration
+	ns             string
+}
+
+func (c *HPAWatcherClient) hpaWatcher(res string) (new []*v2beta1.HorizontalPodAutoscaler, modified []*v2beta1.HorizontalPodAutoscaler, deleted []*v2beta1.HorizontalPodAutoscaler, resVer string, err error) {
+	hpaInterface := c.clientSet.AutoscalingV2beta1()
+
+	metaOptions := metav1.ListOptions{Watch: true, ResourceVersion: res}
+
+	hpaWatcher, err := hpaInterface.HorizontalPodAutoscalers(metav1.NamespaceAll).Watch(metaOptions)
+	if err != nil {
+		log.Infof("Failed to watch %v: %v", expectedHPAType, err)
+	}
+	defer hpaWatcher.Stop()
+
+	watcherTimeout := time.NewTimer(c.readTimeout)
+	for {
+		select {
+		case rcvdHPA, ok := <-hpaWatcher.ResultChan():
+			if !ok {
+				log.Debugf("Unexpected watch close")
+				return nil, nil, nil, "0", err
+			}
+			currHPA, ok := rcvdHPA.Object.(*v2beta1.HorizontalPodAutoscaler)
+			if !ok {
+				log.Infof("Wrong type: %s", currHPA)
+				continue
+			}
+			if currHPA.ResourceVersion != "" && currHPA.ResourceVersion != resVer {
+				resVer = currHPA.ResourceVersion
+			}
+			if rcvdHPA.Type == watch.Error {
+				log.Infof("Error in the watcher, evaluating: %s", currHPA)
+				//status, ok := rcvdHPA.Object.()
+			}
+			if rcvdHPA.Type == watch.Added {
+				log.Infof("Adding this manifest: %s", currHPA)
+				new = append(new, currHPA)
+			}
+			if rcvdHPA.Type == watch.Modified {
+				log.Infof("Modifying this manifest: %s", currHPA)
+				modified = append(modified, currHPA)
+			}
+			if rcvdHPA.Type == watch.Deleted {
+				deleted = append(deleted, currHPA)
+			}
+
+			watcherTimeout.Reset(c.readTimeout)
+		case <-watcherTimeout.C:
+			return new, modified, deleted, resVer, nil
+		}
+	}
+}
+
+func newHPAWatcher() (*HPAWatcherClient, error) {
+	namespace := as.GetResourcesNamespace()
+	clientAPI, err := as.GetAPIClient()
+	if err != nil {
+		log.Errorf("Error creating Client for the HPA: %s", err.Error())
+		return nil, err
+	}
+	hpaPollItl := config.Datadog.GetInt("hpa_watcher_polling_freq")
+	hpaRefreshItl := config.Datadog.GetInt("hpa_external_metrics_polling_freq")
+	hpaExternalMaxAge := config.Datadog.GetInt("hpa_external_metrics_max_age")
```

Comment:
These local variables don't really need the `hpa` prefix either. Also not sure how obvious `Itl` is as a short for for `Interval`.